### PR TITLE
container: Set `container_engine_t` SELinux label

### LINKED ISSUE #880: SELinux in enforcing mode breaks nested gVisor container
# Problem

While debugging a Dangerzone 0.7.0 bug report on a Fedora 40 system (#871), we found out that SELinux in enforcing mode breaks the **inner** gVisor container that runs within the **outer** Podman container.

## Original error

```
$ dangerzone-cli ~/Downloads/sample.pdf
[...]
[DEBUG] Marking doc qyU8c4 as 'converting'
[INFO] > /usr/bin/podman run --log-driver none --security-opt no-new-privileges --cap-drop all --cap-add SYS_CHROOT --network=none -u dangerzone --rm -i --name dangerzone-doc-to-pixels-qyU8c4 --userns nomap dangerzone.rocks/dangerzone /usr/bin/python3 -m dangerzone.conversion.doc_to_pixels
[ERROR] [doc qyU8c4] 0% Unspecified error
[DEBUG] Marking doc qyU8c4 as 'failed'
```

## Output from Podman container with gVisor debug logging enabled

```
$ /usr/bin/podman run -e RUNSC_DEBUG=1 --log-driver none --security-opt no-new-privileges --cap-drop all --cap-add SYS_CHROOT --network=none -u dangerzone --rm -i --name dangerzone-doc-to-pixels-qyU8c4 --userns nomap dangerzone.rocks/dangerzone /usr/bin/python3 -m dangerzone.conversion.doc_to_pixels

8< ... snip ... >8

I0725 14:19:04.811776      20 main.go:197] **************** gVisor ****************
I0725 14:19:04.812445      20 boot.go:258] Setting product_name: "Standard PC (Q35 + ICH9, 2009)"
W0725 14:19:04.813081      20 specutils.go:129] noNewPrivileges ignored. PR_SET_NO_NEW_PRIVS is assumed to always be set.
I0725 14:19:04.813105      20 chroot.go:91] Setting up sandbox chroot in "/tmp"
W0725 14:19:04.813312      20 chroot.go:108] Failed to copy /etc/localtime: open /etc/localtime: no such file or directory. UTC timezone will be used.
I0725 14:19:04.813335      20 chroot.go:36] Mounting "/proc" at "/tmp/proc"
W0725 14:19:04.813394      20 util.go:64] FATAL ERROR: error setting up chroot: error remounting chroot in read-only: permission denied
error setting up chroot: error remounting chroot in read-only: permission denied
D0725 14:19:04.814640       7 sandbox.go:1278] Destroying sandbox "dangerzone"
D0725 14:19:04.814726       7 sandbox.go:1287] Killing sandbox "dangerzone"
D0725 14:19:04.814809       7 container.go:776] Destroy container, cid: dangerzone
D0725 14:19:04.814847       7 container.go:1087] Killing gofer for container, cid: dangerzone, PID: 15
W0725 14:19:04.814928       7 util.go:64] FATAL ERROR: running container: creating container: cannot create sandbox: cannot read client sync file: waiting for sandbox to start: EOF
running container: creating container: cannot create sandbox: cannot read client sync file: waiting for sandbox to start: EOF
W0725 14:19:04.815004       7 main.go:227] Failure to execute command, err: 1
gVisor quit with exit code: 128
```

## SELinux denial taken from audit logs

> Jul 25 17:51:56 10.0.2.15 audit[8164]: AVC avc:  denied  { mounton } for  pid=8164 comm="exe" path="/proc/fs" dev="proc" ino=4026531847 scontext=system_u:system_r:container_t:s0:c139,c177 tcontext=system_u:object_r:proc_t:s0 tclass=dir permissive=0

# Workarounds

We are currently aware of the following two workarounds:
1. Set SELinux to **permissive** mode: `sudo setenforce Permissive`
2. Disable SELinux labeling for the spawned Podman container with `--security-opt label=disable` (requires changes to Dangerzone code).

### PR DESCRIPTION
**Pull request by apyrgio**
Set the `container_engine_t` SELinux on the **outer** Podman container, so that gVisor does not break on systems where SELinux is enforcing. This label is provided for container engines running within a container, which fits our `runsc` within `crun` situation.

We have considered using the more permissive `label=disable` option, to disable SELinux labels altogether, but we want to take advantage of as many SELinux protections as we can, even for the **outer** container.

Fixes #880

### CODE REVIEW

**Comment by rocodes** [comment]
(Will step through basic Fedora test plan today per @apyrgio: install current, confirm bug, build from this branch, confirm successful conversion)

**Comment by rocodes** [comment]
### Test plan
environment: fedora 40 / dz 0.7.0 
- [x] Install Dangerzone 0.7.0
- [x] Attempt to convert a file and confirm https://github.com/freedomofpress/dangerzone/issues/871 ("unspecified error", document fails to convert)
- [x] Build from source at https://github.com/freedomofpress/dangerzone/pull/881/commits/e1e63d14f8d7d1b6b14b56a86232eef9555ac9d6 following dev docs
- [x] Conversion succeeds

**Comment by apyrgio** [comment]
Thanks a lot Ro for reviewing this PR. Do tell me btw if the rationale makes sense to you. To the best of my knowledge, labeling the outer container with `container_engine_t` is not _bad_, but I haven't seen lots of folks doing that.

In the meantime, I'll do two things:
1. Test on a Windows machine as well, since I have tested only on macOS and Debian-derivatives.
2. See if I can add a test where SELinux is enforcing. It will not be for a Fedora runner, since GitHub does not allow this (see https://github.com/actions/runner-images/issues/2307), but it will get us closer to the end goal.

**Comment by apyrgio** [comment]
> See if I can add a test where SELinux is enforcing. It will not be for a Fedora runner, since GitHub does not allow this (see https://github.com/actions/runner-images/issues/2307), but it will get us closer to the end goal.

This is not as simple as I originally thought. I looked for uses of SELinux in other projects' GitHub actions (see this [GitHub search](https://github.com/search?q=path%3A.github%2Fworkflows%2F*.yml+selinux&type=code&ref=advsearch)), and I found two main ways to work with SELinux enabled; either by starting VMs (lots of Vangrant invocations),  or in self-hosted runners.

Still, I tried to install SELinux in an Ubuntu environment and enable it, but it requires a system reboot to do so. I tried to reboot the `kexec` way, but that did not work as well, at which point I ran out of ideas.

**Comment by apyrgio** [comment]
> Test on a Windows machine as well, since I have tested only on macOS and Debian-derivatives.

Tested the PR on a Windows machine, and it works :tada:

**Comment by rocodes** [comment]
Thanks. Based on that and additional external review feedback you've had on this, happy to approve if you can nail down macOS testing, or if you need that from us just let me know :) 

I'm leaning heavily on the other feedback you've had re: the approach itself. It looks sensible and from what I understand based on upstream info (https://github.com/containers/container-selinux/blob/main/container_contexts, https://github.com/containers/container-selinux/blob/a68865582e123856c191fe0ecbbba9301758e591/container.te#L1423-L1426) and some other documentation sources, it looks right, but I don't consider myself a qualified reviewer on the container side of things so I'm leaning on some  the out of band feedback you've shared with me.

**Review by rocodes — APPROVED** [review]
@apyrgio I'm approving based on your testing + feedback, my testing, and CI for macOS, but leaving unmerged: I'll leave it to you to either conduct further macOS testing on hardware if you feel you need, or to go ahead and merge if you're trying to get the hotfix out the door, so that we aren't holding up your release mechanics.

**Comment by apyrgio** [comment]
Quick update here, we've successfully tested on all supported platforms, and therefore we've merged the PR. Note that for the 0.7.0 hotfix release, we aim to publish just the new Fedora packages (actually, cherry-pick this commit to a `hotfix-0.7.0` branch) and merely bump their patch level to `2`.

Still, it's important to know that our approach works in other platforms, hence the extended review.

**Pull request merged**
